### ISSUE TITLE
Pinned versions in setup.py causes dependency problems

### ISSUE THREAD

**Issue by csande** [snowflakedb/snowflake-connector-python#13]
It is not considered to be good practice to pin versions in the install_requires section of setup.py:

https://packaging.python.org/requirements/?highlight=install_requires

Pinned versions are best left to requirements.txt files for the purpose of achieving repeatable installations. The install_requires option is used to specify "abstract" dependencies.

This lack of flexibility is causing me some problems because other packages I use have different version requirements for packages like boto3.

If it is desired to place some limitations on the versions of packages that are depended upon, it is better to use either >= and/or >= to specify min/max or ranges. Pinning to specific versions is far too restrictive for install_requires.

Please update the setup.py to fix this so that developers that are using your package can have more control over their installations.

**Comment by smtakeda** [comment]
Yes, it is known area that I want to improve. I understand that loosing the version restriction of dependencies will yield flexibility of application's choice. In that sense, I admit that we don't follow the best practice as of today.

In Snowflake end, a challenge is it would need more tests with different combinations of dependencies to ensure the driver works. In particular, `botocore` are somewhat sensitive where the vendored  `requests` is monkey patched to enable OCSP validation. If `requests` is changed from 2.7.0, I cannot guarantee it will work. Fortunately `botocore` has not changed it for a long time, so most versions of `botocore` should be ok, though.

Alternatively we could add `boto3` and `botocore` to vendor folder to isolate the namespace, but it appeared it didn't work well as some reference of module within boto is absolute name, so it cannot be wrapped in a private namespaces.

Anyway let us try improving this issue. `boto3` and `botocore` specific issue could be addressed shortly as we improve CI. `cryptography` dependency is another story.

Thanks for the feedback.

**Comment by csande** [comment]
Thanks for the quick reply. I'll keep an eye on this thread for updates.

**Comment by blr246** [comment]
+1 this issue is interfering majorly with my dependency management tooling. Going to have to make a special workaround for now.

**Comment by smtakeda** [comment]
wip

**Comment by blr246** [comment]
Thank you @smtakeda this is a big help! I am pretty confident this will unblock me. I had made a fork for now to do basically the same thing you did with the boto requirements.

**Comment by smtakeda** [comment]
scheduled the release 1.3.19 this Thursday.

**Comment by csande** [comment]
This looks good and was basically was I was looking for. I also created a fork in order to get around this issue, and I expect I will be able to cease using the fork now that this change has been made.

**Comment by atrigent** [comment]
There's still an issue with the `cryptography` and `pyOpenSSL` dependencies. With `pyOpenSSL>=16.2.0`, currently version `17.1.0` will be installed. However, this depends on `cryptography>=1.9`, which conflicts with `cryptography>=1.7,<1.8` in the snowflake connector.

**Comment by smtakeda** [comment]
ah, ok @atrigent. I'm making further changes. running the internal tests.

**Comment by smtakeda** [comment]
released 1.4.0 (skipped 1.3.19)
Try and let me know.

**Comment by csande** [comment]
The new setup.py again contains pinned versions. If there is a concern about the latest version of pyasn1, can you please use <= or < instead of ==?
